grpc.go

```diff
@@ -623,6 +627,97 @@ func (a *agentGRPC) WaitProcess(ctx context.Context, req *pb.WaitProcessRequest)
 	}, nil
 }
 
+func getPIDIndex(title string) int {
+	// looking for PID field in ps title
+	fields := strings.Fields(title)
+	for i, f := range fields {
+		if f == "PID" {
+			return i
+		}
+	}
+	return -1
+}
+
+func (a *agentGRPC) ListProcesses(ctx context.Context, req *pb.ListProcessesRequest) (*pb.ListProcessesResponse, error) {
+	resp := &pb.ListProcessesResponse{}
+
+	c, err := a.sandbox.getContainer(req.ContainerId)
+	if err != nil {
+		return resp, err
+	}
+
+	// Get the list of processes that are running inside the containers.
+	// the PIDs match with the system PIDs, not with container's namespace
+	pids, err := c.container.Processes()
+	if err != nil {
+		return resp, err
+	}
+
+	switch req.Format {
+	case "table":
+	case "json":
+		resp.ProcessList, err = json.Marshal(pids)
+		return resp, err
+	default:
+		return resp, fmt.Errorf("invalid format option")
+	}
+
+	psArgs := req.Args
+	if len(psArgs) == 0 {
+		psArgs = []string{"-ef"}
+	}
+
+	// All container's processes are visibles from agent's namespace.
+	// pids already contains the list of processes that are running
+	// inside a container, now we have to use that list to filter
+	// ps output and return just container's processes
+	cmd := exec.Command("ps", psArgs...)
```

Comment:
Agree with gwhaley. Also, we should really avoid execing if possible.

Reply:
see https://github.com/opencontainers/runc/blob/master/ps.go#L63

the idea behind runtime's `ps` sub-command is to list the processes using linux [ps](https://linux.die.net/man/1/ps) command,  _"peruse the /proc"_ is not an option since the user decides the output format and the information to show. Would be nice to have an implementation of [ps](https://linux.die.net/man/1/ps) in `go` to include it here, but as far as  I know there is no. If you concern is that we don't include `ps` command in the image, don't worries I already filed an [issue](https://github.com/kata-containers/osbuilder/issues/81) for that

Reply:
Ah, so the user can supply arguments to `ps` from the docker command line? In which case, yes, it makes more sense to use `ps` directly.
I will note, you might find that some implementations of `ps` (notably busybox, which then also implies Alpine I believe), may act differently and/or not support all options that you get with the normal 'full ps'. So, you may wish to beware and test that out.